org.bndtools.builder/src/org/bndtools/builder/classpath/BndContainerCompilationParticipant.java

```diff
@@ -5,7 +5,10 @@
 import org.eclipse.core.runtime.CoreException;
 import org.eclipse.jdt.core.IJavaProject;
 import org.eclipse.jdt.core.compiler.CompilationParticipant;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.ServiceScope;
 
+@Component(scope = ServiceScope.PROTOTYPE, service = CompilationParticipant.class)
```

Comment:
We might want to put comments in the places where the scope is set to explain why we need that scope for the component. There is a mix of scope types.

Reply:
Eclipse registry extension points generally assume that a new instance is returned every time it is instantiated. So unless you know it is safe to do otherwise, you must use PROTOTYPE scope.

When it is safe to use SINGLETON scope: if the base class is an interface, and your implementation doesn't maintain any per-instance state, then you can sometimes get away with SINGLETON. But in this case, the base service type `CompilationParticipant` is a class and so it might maintain some state of its own. So I went with the PROTOTYPE as the least likely to cause problems. To be honest, I haven't looked much further than this and it's *possible* that some of the ones I've set as PROTOTYPE could be SINGLETON.

Reply:
Sorry, I just had a closer look at `CompilationParticipant` and realised that it is stateless. So now I understand your confusion as to why it was marked PROTOTYPE.

I think, in my rush to try and get something working, I set everything to PROTOTYPE as the safest option, with the thought that I would at some point return to double-check and change some to SINGLETON if they were required. I'll tidy this up.